cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java

```diff
@@ -38,8 +42,10 @@
   private static final long serialVersionUID = -1173500180640174909L;
 
   private Map<String, Map<String, Long>> mapReduceCounters;
-  private final Map<Scope, Map<String, List<NodeValue>>> tokenValueMap = new EnumMap<>(Scope.class);
+  private Map<Scope, Map<String, List<NodeValue>>> tokenValueMap =
+    Collections.synchronizedMap(new EnumMap<Scope, Map<String, List<NodeValue>>>(Scope.class));
```

Comment:
And in fact, from the current logic, the top level map doesn't need to be synchronized since you put the values in the constructor and never add new scopes afterward.